core/src/main/java/org/apache/iceberg/IncrementalDataScan.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Sets;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+
+class IncrementalDataScan extends DataTableScan implements TableScan {
+  private final TableOperations ops;
+  private final Table table;
+  private boolean colStats;
+  private long fromSnapshotId;
+  private long toSnapshotId;
+  private Set<Long> snapshotIds;
+
+
+  IncrementalDataScan(TableOperations ops, Table table, long fromSnapshotId, long toSnapshotId) {
+    super(ops, table);
+    this.ops = ops;
+    this.table = table;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+    this.snapshotIds = snapshotsWithin(table, fromSnapshotId, toSnapshotId);
+  }
+
+  /**
+   * @return all snapshots ids between {@param fromSnapshotId } exclusive
+   * and {@param toSnapshotId} inclusive
+   */
+  private static Set<Long> snapshotsWithin(Table table, long fromSnapshotId, long toSnapshotId) {
```

Comment:
I don't think it is correct to use table history for this. Table history is a log of changes to the current snapshot ID.

When a table is rolled back, the current snapshot ID is set explicitly. So the snapshot log might go from snapshot A, to B, to C, back to B, and then to D. Using the history table for the snapshot range A to D would result in processing C, even though it was rolled back, and processing B twice.

Also, the snapshot log can be missing entries when transactions are used. Transactions are are implemented by staging changes without committing, and then committing the final result and the latest table state. If you had a transaction of two append operations based on snapshot A, the table would contain snapshots A, B, and C, where the parent of C is B and the parent of B is A. But the history log will only show that the current snapshot went from A to C.

I think a better way to get changes is to use the parent relationship between snapshots. That way, if you have history A, to C, to D, to C, you'd get C, B, and A. [`SnapshotUtil`](https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L28-L58) has some code that traverses ancestors.

Also, it is critical that you skip `replace` snapshots because the data has not changed, but grouping in files has. If you only process added files and skip replace then you'll not process data twice. Also, delete and overwrite should probably throw an exception?

Reply:
That makes sense!. I'll change this to use SnapshotUtil

Reply:
Fixed to used `SnapshotUtil`